### ISSUE TITLE
setdisplay does not work as well as Gadly plots on Media master

### ISSUE THREAD

**Issue by rohitvarkey** [JunoLab/Media.jl#3]
``` julia
julia> Pkg.checkout("Media")
INFO: Checking out Media master...
INFO: Pulling Media latest master...
INFO: No packages to install, update or remove

julia> using DevTools, Gadfly
INFO: Precompiling module Media...

julia> plot(x = 1:100, y = cumsum(rand(100)-0.5), Geom.line) #opens the browser to display rather than the nice Blink window display

julia> using Media

julia> setdisplay(Media.Tabular, BlinkDisplay._display)
DevTools.BlinkDisplay.WebView(Any[],nothing)

julia> rand(5, 5) #output in REPL rather than the Blink window
5x5 Array{Float64,2}:
 0.572228  0.691744  0.633762  0.30069   0.31797
 0.415801  0.120471  0.194096  0.771513  0.26699
 0.663217  0.951212  0.749737  0.620215  0.992151
 0.351145  0.60946   0.440976  0.312864  0.834909
 0.540536  0.515108  0.944356  0.576629  0.923103

julia> setdisplay(FloatingPoint, BlinkDisplay._display)
WARNING: Base.FloatingPoint is deprecated, use AbstractFloat instead.
  likely near no file:0
DevTools.BlinkDisplay.WebView(Any[],nothing)

julia> 2.3 #output in the REPL again.
2.3
```

Could this be related to precompilation? I tried cleaning out my `libs`  and then recompiling without any luck.

**Comment by MikeInnes** [comment]
Hmm, if it was a precompilation issue I'd expect it to happen the other way around, if anything. Will take a  look.

**Comment by hustf** [comment]
I reproduced rohitvarkey's behaviour. Then, to try and isolate the issue

```
julia>using Gadfly
```

...followed by running through system.jl outside of the Media module (leaving out only module Media and Media.) until I got the first error message at compat.jl:20

```
julia> pushdisplay(DisplayHook())
Error showing value of type Array{Display,1}:
ERROR: ArgumentError: reducing over an empty collection is not allowed
 in mapfoldl at reduce.jl:58
 in nearest at none:1
 in getdisplay at none:3
 in display at none:1
 [inlined code] from multimedia.jl:151
 in display at multimedia.jl:162
 in print_response at REPL.jl:134
 in print_response at REPL.jl:121
 in anonymous at REPL.jl:624
 in run_interface at LineEdit.jl:1610
 in run_frontend at REPL.jl:863
 in run_repl at REPL.jl:167
 in _start at client.jl:420
```

Right before running line 20, I checked:

```
julia> methods(display)
# 17 methods for generic function "display":
display(d::Display, mime::AbstractString, x) at multimedia.jl:110
display(mime::AbstractString, x) at multimedia.jl:111
display(d::TextDisplay, M::MIME{symbol("text/plain")}, x) at multimedia.jl:119
display(d::TextDisplay, x) at multimedia.jl:120
display(m::MIME{mime}, x) at multimedia.jl:170
display(d::Base.REPL.REPLDisplay{R<:Base.REPL.AbstractREPL}, ::MIME{symbol("text
/plain")}, x) at REPL.jl:112
display(d::Base.REPL.REPLDisplay{R<:Base.REPL.AbstractREPL}, md::Base.Markdown.M
D) at markdown/render/terminal/render.jl:126
display(d::Base.REPL.REPLDisplay{R<:Base.REPL.AbstractREPL}, md::Array{Base.Mark
down.MD,1}) at markdown/Markdown.jl:61
display(d::Base.REPL.REPLDisplay{R<:Base.REPL.AbstractREPL}, ::MIME{symbol("imag
e/png")}, p::Gadfly.Plot) at C:\Users\F\.julia\v0.4\Gadfly\src\Gadfly.jl:952
display(d::Base.REPL.REPLDisplay{R<:Base.REPL.AbstractREPL}, ::MIME{symbol("imag
e/svg+xml")}, p::Gadfly.Plot) at C:\Users\F\.julia\v0.4\Gadfly\src\Gadfly.jl:961

display(d::Base.REPL.REPLDisplay{R<:Base.REPL.AbstractREPL}, ::MIME{symbol("text
/html")}, p::Gadfly.Plot) at C:\Users\F\.julia\v0.4\Gadfly\src\Gadfly.jl:970
display(d::Base.REPL.REPLDisplay{R<:Base.REPL.AbstractREPL}, ::MIME{symbol("appl
ication/postscript")}, p::Gadfly.Plot) at C:\Users\F\.julia\v0.4\Gadfly\src\Gadf
ly.jl:1003
display(d::Base.REPL.REPLDisplay{R<:Base.REPL.AbstractREPL}, ::MIME{symbol("appl
ication/pdf")}, p::Gadfly.Plot) at C:\Users\F\.julia\v0.4\Gadfly\src\Gadfly.jl:1
012
display(d::Base.REPL.REPLDisplay{R<:Base.REPL.AbstractREPL}, x) at REPL.jl:117
display(p::Compose.Context) at C:\Users\F\.julia\v0.4\Compose\src\Compose.jl:222

display(p::Gadfly.Plot) at C:\Users\F\.julia\v0.4\Gadfly\src\Gadfly.jl:922
display(x) at multimedia.jl:162
```

Right after running line 20, I checked again:

```
julia> methods(display)
Error showing value of type MethodTable:
ERROR: ArgumentError: reducing over an empty collection is not allowed
 in mapfoldl at reduce.jl:58
 in nearest at none:1
 in getdisplay at none:3
 in display at none:1
 [inlined code] from multimedia.jl:151
 in display at multimedia.jl:162
 in print_response at REPL.jl:134
 in print_response at REPL.jl:121
 in anonymous at REPL.jl:624
 in run_interface at LineEdit.jl:1610
 in run_frontend at REPL.jl:863
 in run_repl at REPL.jl:167
 in _start at client.jl:420
```

I guess the most probable cause is me having left out a line somewhere? If it works on other systems?

**Comment by rohitvarkey** [comment]
@hustf I'm not sure if that is relevant to this issue. This issue doesn't occur on Media `v0.1.1` and if you look at the commits to master after that, they are not very significant changes (except for enabling precompilation). You can examine them [here](https://github.com/JunoLab/Media.jl/compare/v0.1.1...master). As you'd notice `compat.jl` hasn't changed at all!

**Comment by hustf** [comment]
I can not reason against that, Rohitvarkey.

What else has changed on Master? Could some other package influence the behaviour? I thought it might be interesting to add a confirmation. The below indicates changes to Media in master is the cause: Everything works.

You might want to refer to the report just now on https://github.com/JunoLab/Blink.jl/issues/42. Continuing from the last point there:

5)

```
julia> Pkg.status("Media")
 - Media                         0.1.1

julia> using Media

julia> setdisplay(Media.Tabular, BlinkDisplay._display)
DevTools.BlinkDisplay.WebView(Any[Blink.AtomShell.Window(2,Blink.AtomShell.Shel
link\deps\atom\electron.exe' 'C:\Users\F\.julia\v0.4\Blink\src\AtomShell\main.j
cket(active, 0 bytes waiting),Dict{ASCIIString,Any}("callback"=>(anonymous func
ocket(5,TCPSocket(closed, 0 bytes waiting),false,false),Dict{ASCIIString,Any}("
dition(Any[])))],nothing)

julia> rand(5, 5)
```

(the matrix appears in a Chrome Window)

6)

```
julia> setdisplay(FloatingPoint, BlinkDisplay._display)
WARNING: Base.FloatingPoint is deprecated, use AbstractFloat instead.
  likely near no file:0
DevTools.BlinkDisplay.WebView(Any[Blink.AtomShell.Window(2,Blink.AtomShell.Shell(Process(`'C:\Users\F\.julia\v0.4\B
link\deps\atom\electron.exe' 'C:\Users\F\.julia\v0.4\Blink\src\AtomShell\main.js' port 7434`, ProcessRunning),TCPSo
cket(active, 0 bytes waiting),Dict{ASCIIString,Any}("callback"=>(anonymous function))),Blink.Page(1,WebSockets.WebS
ocket(5,TCPSocket(closed, 0 bytes waiting),false,false),Dict{ASCIIString,Any}("callback"=>(anonymous function)),Con
dition(Any[]))),Blink.AtomShell.Window(3,Blink.AtomShell.Shell(Process(`'C:\Users\F\.julia\v0.4\Blink\deps\atom\ele
ctron.exe' 'C:\Users\F\.julia\v0.4\Blink\src\AtomShell\main.js' port 7434`, ProcessRunning),TCPSocket(active, 0 byt
es waiting),Dict{ASCIIString,Any}("callback"=>(anonymous function))),Blink.Page(2,WebSockets.WebSocket(9,TCPSocket(
active, 0 bytes waiting),false,false),Dict{ASCIIString,Any}("callback"=>(anonymous function)),Condition(Any[])))],n
othing)

julia> 1+1.3
```

(2.3 appears in a Chrome window)

**Comment by MikeInnes** [comment]
Is this fixed for you?

**Comment by hustf** [comment]
On Media master it seems to work now. 

When recompiling, there are some warnings from Reactive which will perhaps be causing trouble when using other  packages: Reactive map(Any, Reactive.Signal) is ambiguous with two functions in Lazy: map(Function, Lazy.List...) and map(DataType, Lazy.List...). Isn`t Atom using those Lazy functions too?

I believe I got the same warnings yesterday, when using Reactive through GLWindow. 

I did some experiments with displaying Compose graphics and SVG graphics through the same interface, but every time the output goes to the REPL I need to take a long break. Perhaps that could be a nice example to put in the docs? I gave up after a few tries.
